dotnet/src/Connectors/Connectors.Memory.Postgres/PostgresVectorStoreRecordCollection.cs

```diff
@@ -0,0 +1,317 @@
+    }
+
+    /// <inheritdoc />
+    public Task<VectorSearchResults<TRecord>> VectorizedSearchAsync<TVector>(TVector vector, VectorSearchOptions? options = null, CancellationToken cancellationToken = default)
+    {
+        Verify.NotNull(vector);
+
+        var vectorType = vector.GetType();
+
+        if (!PostgresConstants.SupportedVectorTypes.Contains(vectorType))
+        {
+            throw new NotSupportedException(
+                $"The provided vector type {vectorType.FullName} is not supported by the SQLite connector. " +
+                $"Supported types are: {string.Join(", ", PostgresConstants.SupportedVectorTypes.Select(l => l.FullName))}");
+        }
+
+        var searchOptions = options ?? s_defaultVectorSearchOptions;
+        var vectorProperty = this.GetVectorPropertyForSearch(searchOptions.VectorPropertyName);
+
+        if (vectorProperty is null)
+        {
+            throw new InvalidOperationException("The collection does not have any vector properties, so vector search is not possible.");
+        }
+
+        var pgVector = PostgresVectorStoreRecordPropertyMapping.MapVectorForStorageModel(vector);
+
+        Verify.NotNull(pgVector);
+
+        // Simulating skip/offset logic locally, since OFFSET can work only with LIMIT in combination
+        // and LIMIT is not supported in vector search extension, instead of LIMIT - "k" parameter is used.
+        var limit = searchOptions.Top + searchOptions.Skip;
+
+        var results = this._client.GetNearestMatchesAsync(
+            this.CollectionName,
+            this._propertyReader.RecordDefinition.Properties,
+            vectorProperty,
+            pgVector,
+            searchOptions.Top,
+            searchOptions.Filter,
+            searchOptions.Skip,
+            searchOptions.IncludeVectors,
+            cancellationToken
+            ).Select(result =>
+            {
+                var record = this._mapper.MapFromStorageToDataModel(
+                    result.Row, new StorageToDataModelMapperOptions() { IncludeVectors = searchOptions.IncludeVectors });
+
+                return new VectorSearchResult<TRecord>(record, result.Distance);
+            }, cancellationToken);
+
+        return Task.FromResult(new VectorSearchResults<TRecord>(results));
+    }
+
+    private async Task InternalCreateCollectionAsync(bool ifNotExists, CancellationToken cancellationToken = default)
+    {
+        await this._client.CreateTableAsync(this.CollectionName, this._propertyReader.RecordDefinition, ifNotExists, cancellationToken).ConfigureAwait(false);
+        // Create indexes for vector properties.
+        foreach (var vectorProperty in this._propertyReader.VectorProperties)
+        {
+            // Ensure the dimensionality of the vector is supported for indexing.
+            if (vectorProperty.IndexKind == IndexKind.Hnsw)
+            {
+                if (vectorProperty.Dimensions > 2000)
+                {
+                    this._logger.LogWarning(
```

Comment:
I think in this case instead of logging a warning, it might be better to throw.  It wouldn't be great to generate embeddings for a lot of records and upsert them, only to realise afterwards that you missed a warning that means that you cannot search any of the data.

Reply:
By the way, we didn't introduce any telemetry in other connectors yet. I think we should think about a solution that will be applicable for all connectors at the same time which will include logging, metering and tracing. So, in case you see that you spend a lot of time on telemetry - I would recommend focusing on the db functionality, because we will definitely work more on telemetry in the future.

Reply:
I agree that an exception is thrown for index creation failure, but the collection table still allowed to be created successfully?

Reply:
Modified to wrap index creation and table creation into a transaction.

Reply:
For general telemetry I was thinking of just adding a decorator class that can be used with any connector.  That way each connector doesn't have to implement it itself.  That's assuming the user wants telemetry, since it might be duplicating what is already logged by the db client.